pkg/envoy/xds/ack.go

```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
 
 		// Completion didn't match or is still waiting for some ACKs. Keep it
 		// in the pending list.
-		remainingCompletions = append(remainingCompletions, comp)
+		remainingCompletions[comp] = pending
+	}
+
+	m.pendingCompletions = remainingCompletions
+}
+
+// HandleResourceVersionNack handles NACKs by calling callbacks of pending completions, if set.
+func (m *AckingResourceMutatorWrapper) HandleResourceVersionNack(version uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string, detail string) {
+	ackLog := log.WithFields(logrus.Fields{
+		logfields.XDSVersionInfo: version,
+		logfields.XDSClientNode:  node,
+		logfields.XDSTypeURL:     typeURL,
+	})
+
+	nodeID, err := m.nodeToID(node)
+	if err != nil {
+		// Ignore NACKs from unknown or misconfigured nodes which have invalid
+		// node identifiers.
+		ackLog.Warning("Invalid ID in Node identifier; ignoring NACK")
+		return
+	}
+
+	m.locker.Lock()
+	defer m.locker.Unlock()
+
+	remainingCompletions := make(map[*completion.Completion]*pendingCompletion, len(m.pendingCompletions))
+
+	for comp, pending := range m.pendingCompletions {
+		if comp.Context().Err() != nil {
+			// Completion was canceled or timed out.
+			// Remove from pending list.
+			ackLog.Debugf("completion context was canceled: %v", pending)
+			comp.Complete(context.Canceled)
+			continue
+		}
+
+		// Check if comp's version was NACKed
+		if pending.version == version && pending.typeURL == typeURL {
```

Comment:
I think it's incorrect to do an `==` test here. If a later version is NACKed, you may never notify a completion's ACK or NACK.
I think you should test `pending.version <= version` like for an ACK. I.e. if a version is NACKed, all previous pending versions are NACKed too.

Reply:
Actually, can you confirm what is the `version` in a NACK? Is it the version of the last version ACKed? If that's the case, shouldn't we ACK all versions <=version?

Reply:
Digging into this revealed a hole I've been digging myself out of the whole day..
To reliably know what was NACKed I had to change Nonce in a response to be the same as VersionInfo. Then in a forthcoming request:

- VersionInfo indicated the last ACKed version
- ResponseNonce, if > VersionInfo, indicates that versions after VersionInfo, up to and including ResponseNonce, were NACKed.